File: 350-intersection-of-two-arrays-ii/intersection-of-two-arrays-ii.py
from collections import Counter

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # count frequencies of elements
        counts = Counter(nums1)
        result = []

        for num in nums2:
            # if the elements is still available in out count pool
            if counts[num] > 0:
                result.append(num)
                counts[num] -= 1  # decrement the count

        return result

File: 350-intersection-of-two-arrays-ii/intersection-of-two-arrays-ii.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> counts;
        vector<int> result;

        // counts frequencies of elements in nums1
        for (int &num : nums1) {
            counts[num]++;
        }

        // match elements from nums2
        for (int &num : nums2) {
            if (counts[num] > 0) {
                result.push_back(num);
                counts[num]--;
            }
        }
        return result;
    }
};